[master] opam init fails when no system compiler is present

With opam master (7d926f1f6479929c6d0f084053c1583a35e48d3c, compiled using make cold), on a clean system, without any ocaml compiler available:
```
[opam@edb63d1833b7 opam]$ ./opam init 
[...]
[WARNING] Shell not updated in non-interactive mode: use --shell-setup
<><> Creating initial switch 'default' (invariant ["ocaml" {>= "4.05.0"}] - initially with ocaml-system) 
<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><><><>
Switch invariant: ["ocaml" {>= "4.05.0"}]
[ERROR] Could not determine which packages to install for this switch:
  * Missing dependency:
    - (request) → ocaml-system
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
Switch initialisation failed: clean up? ('n' will leave the switch partially installed) [Y/n] y
[NOTE] Opam has been initialised, but the initial switch creation failed.
       Use 'opam switch create <compiler>' to get started.
```

--- a/master_changes.md
+++ b/master_changes.md
@@ -17,6 +17,7 @@ New option/command/subcommand are prefixed with ◈.
 ## Init
   * Introduce a `default-invariant` config field, restore the 2.0 semantics for
     `default-compiler` [#4607 @AltGr]
+  * Fix default invariant with no system compiler [#4644 @AltGr - fix #4640]
 
 ## Config report
   * Fix `Not_found` (config file) error [#4570 @rjbou]

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -902,7 +902,9 @@ let init
               (OpamFile.InitConfig.default_compiler init_config)
           in
           let invariant = OpamFile.InitConfig.default_invariant init_config in
-          let virt_st = OpamSwitchState.load_virtual gt rt in
+          let virt_st =
+            OpamSwitchState.load_virtual ~avail_default:false gt rt
+          in
           let univ =
             OpamSwitchState.universe virt_st
               ~requested:OpamPackage.Name.Set.empty Query

--- a/tests/reftests/init.test
+++ b/tests/reftests/init.test
@@ -10,20 +10,21 @@ Configuring from ${BASEDIR}/opamrc and then from built-in defaults.
 <><> Fetching repository information ><><><><><><><><><><><><><><><><><><><><><>
 [default] Initialised
 
-<><> Creating initial switch 'default' (invariant ["ocaml" {>= "4.05.0"}] - initially with ocaml-system) 
+<><> Creating initial switch 'default' (invariant ["ocaml" {>= "4.05.0"}] - initially with ocaml-base-compiler) 
 
 <><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><><><>
 Switch invariant: ["ocaml" {>= "4.05.0"}]
-[ERROR] Could not determine which packages to install for this switch:
-  * Missing dependency:
-    - (request) -> ocaml-system
-    unmet availability conditions, e.g. 'sys-ocaml-version = "4.10.0"'
 
-
-Switch initialisation failed: clean up? ('n' will leave the switch partially installed) [Y/n] y
-[NOTE] Opam has been initialised, but the initial switch creation failed.
-       Use 'opam switch create <compiler>' to get started.
-# Return code 20 #
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+Faking installation of base-bigarray.base
+Faking installation of base-threads.base
+Faking installation of base-unix.base
+Faking installation of ocaml-base-compiler.4.10.0
+Faking installation of ocaml-config.1
+Faking installation of ocaml.4.10.0
+Done.
+### opam switch invariant
+["ocaml" {>= "4.05.0"}]
 ### rm -rf ${OPAMROOT}
 ### <opamrc>
 eval-variables: [ sys-ocaml-version ["echo" "4.02.3"] "old system compiler" ]

--- a/tests/reftests/init.test
+++ b/tests/reftests/init.test
@@ -2,9 +2,33 @@
 ### OPAMYES=1
 ### tar xzf ${OPAMROOT}/repo/default.tar.gz
 ### rm -rf ${OPAMROOT}
-### OPAMVAR_sys_ocaml_version=''
-### opam init --no-setup --bypass-checks default default/ --fake
-No configuration file found, using built-in defaults.
+### <opamrc>
+eval-variables: [ sys-ocaml-version ["false"] "no system compiler" ]
+### opam init --no-setup --bypass-checks default default/ --fake --config opamrc
+Configuring from ${BASEDIR}/opamrc and then from built-in defaults.
+
+<><> Fetching repository information ><><><><><><><><><><><><><><><><><><><><><>
+[default] Initialised
+
+<><> Creating initial switch 'default' (invariant ["ocaml" {>= "4.05.0"}] - initially with ocaml-system) 
+
+<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><><><>
+Switch invariant: ["ocaml" {>= "4.05.0"}]
+[ERROR] Could not determine which packages to install for this switch:
+  * Missing dependency:
+    - (request) -> ocaml-system
+    unmet availability conditions, e.g. 'sys-ocaml-version = "4.10.0"'
+
+
+Switch initialisation failed: clean up? ('n' will leave the switch partially installed) [Y/n] y
+[NOTE] Opam has been initialised, but the initial switch creation failed.
+       Use 'opam switch create <compiler>' to get started.
+# Return code 20 #
+### rm -rf ${OPAMROOT}
+### <opamrc>
+eval-variables: [ sys-ocaml-version ["echo" "4.02.3"] "old system compiler" ]
+### opam init --no-setup --bypass-checks default default/ --fake --config opamrc
+Configuring from ${BASEDIR}/opamrc and then from built-in defaults.
 
 <><> Fetching repository information ><><><><><><><><><><><><><><><><><><><><><>
 [default] Initialised
@@ -25,32 +49,10 @@ Done.
 ### opam switch invariant
 ["ocaml" {>= "4.05.0"}]
 ### rm -rf ${OPAMROOT}
-### OPAMVAR_sys_ocaml_version=4.02.3
-### opam init --no-setup --bypass-checks default default/ --fake
-No configuration file found, using built-in defaults.
-
-<><> Fetching repository information ><><><><><><><><><><><><><><><><><><><><><>
-[default] Initialised
-
-<><> Creating initial switch 'default' (invariant ["ocaml" {>= "4.05.0"}] - initially with ocaml-base-compiler) 
-
-<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><><><>
-Switch invariant: ["ocaml" {>= "4.05.0"}]
-
-<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-Faking installation of base-bigarray.base
-Faking installation of base-threads.base
-Faking installation of base-unix.base
-Faking installation of ocaml-base-compiler.4.10.0
-Faking installation of ocaml-config.1
-Faking installation of ocaml.4.10.0
-Done.
-### opam switch invariant
-["ocaml" {>= "4.05.0"}]
-### rm -rf ${OPAMROOT}
-### OPAMVAR_sys_ocaml_version=4.07.0
-### opam init --no-setup --bypass-checks default default/ --fake
-No configuration file found, using built-in defaults.
+### <opamrc>
+eval-variables: [ sys-ocaml-version ["echo" "4.07.0"] "new system compiler" ]
+### opam init --no-setup --bypass-checks default default/ --fake --config opamrc
+Configuring from ${BASEDIR}/opamrc and then from built-in defaults.
 
 <><> Fetching repository information ><><><><><><><><><><><><><><><><><><><><><>
 [default] Initialised

